File: crates/nil-core/src/infrastructure/building/prefecture.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

mod build_queue;
mod catalog;

use super::{BuildingId, BuildingLevel};
use crate::check_total_resource_ratio;
use crate::infrastructure::requirements::InfrastructureRequirements;
use crate::resource::{
  BaseCost,
  BaseCostGrowth,
  MaintenanceRatio,
  ResourceRatio,
  Workforce,
  WorkforceGrowth,
};
use nil_core_macros::Building;
use serde::{Deserialize, Serialize};

pub use build_queue::{
  PrefectureBuildOrder,
  PrefectureBuildOrderId,
  PrefectureBuildOrderKind,
  PrefectureBuildOrderOptions,
  PrefectureBuildOrderStatus,
  PrefectureBuildQueue,
};
pub use catalog::{PrefectureCatalog, PrefectureCatalogEntry, PrefectureCatalogRecipe};

/// Centro logístico da aldeia, responsável pela construção de edifícios.
#[derive(Building, Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Prefecture {
  level: BuildingLevel,
  enabled: bool,
  build_queue: PrefectureBuildQueue,
}

impl Prefecture {
  pub const ID: BuildingId = BuildingId::Prefecture;

  pub const MIN_LEVEL: BuildingLevel = BuildingLevel::new(1);
  pub const MAX_LEVEL: BuildingLevel = BuildingLevel::new(30);

  pub const BASE_COST: BaseCost = BaseCost::new(150_000);
  pub const BASE_COST_GROWTH: BaseCostGrowth = BaseCostGrowth::new(0.2);
  pub const MAINTENANCE_RATIO: MaintenanceRatio = MaintenanceRatio::new(0.005);

  pub const WOOD_RATIO: ResourceRatio = ResourceRatio::new(0.3);
  pub const STONE_RATIO: ResourceRatio = ResourceRatio::new(0.5);
  pub const IRON_RATIO: ResourceRatio = ResourceRatio::new(0.2);

  pub const WORKFORCE: Workforce = Workforce::new(50);
  pub const WORKFORCE_GROWTH: WorkforceGrowth = WorkforceGrowth::new(0.2);

  pub const INFRASTRUCTURE_REQUIREMENTS: InfrastructureRequirements =
    InfrastructureRequirements::none();

  pub(crate) fn build_queue_mut(&mut self) -> &mut PrefectureBuildQueue {
    &mut self.build_queue
  }

  pub(crate) fn process_queue(&mut self) {
    if self.enabled {
      self.build_queue.process(self.level.into());
    }
  }
}

impl Default for Prefecture {
  fn default() -> Self {
    Self {
      level: BuildingLevel::new(1),
      enabled: true,
      build_queue: PrefectureBuildQueue::default(),
    }
  }
}

check_total_resource_ratio!(
  Prefecture::WOOD_RATIO,
  Prefecture::STONE_RATIO,
  Prefecture::IRON_RATIO,
);

File: app/src/types/core/infrastructure/prefecture.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import type { Coord } from '@/types/core/continent';
import type { Resources } from '@/types/core/resources';
import type { InfrastructureQueueOrder } from '@/types/core/infrastructure/queue';
import type { Infrastructure, InfrastructureRequirements } from '@/types/core/infrastructure';
import type { Building, BuildingId, BuildingLevel } from '@/types/core/infrastructure/building';

export interface Prefecture extends Building {
  readonly buildQueue: PrefectureBuildQueue;
}

export interface PrefectureBuildQueue {
  readonly orders: readonly PrefectureBuildOrder[];
}

export interface PrefectureBuildOrder extends InfrastructureQueueOrder {
  readonly kind: PrefectureBuildOrderKind;
  readonly building: BuildingId;
  readonly level: BuildingLevel;
}

export type PrefectureBuildOrderKind = 'construction' | 'demolition';

export interface PrefectureBuildOrderRequest {
  readonly coord: Coord;
  readonly building: BuildingId;
  readonly kind: PrefectureBuildOrderKind;
}

export type PrefectureBuildCatalog = {
  readonly [B in keyof Infrastructure]: PrefectureBuildCatalogEntry;
};

export type PrefectureBuildCatalogEntry =
  | PrefectureBuildCatalogEntryAvailable
  | PrefectureBuildCatalogEntryMaxed
  | PrefectureBuildCatalogEntryUnmet;

export interface PrefectureBuildCatalogEntryAvailable {
  readonly kind: 'available';
  readonly recipe: PrefectureBuildCatalogRecipe;
}

export interface PrefectureBuildCatalogEntryMaxed {
  readonly kind: 'maxed';
}

export interface PrefectureBuildCatalogEntryUnmet {
  readonly kind: 'unmet';
  readonly requirements: InfrastructureRequirements;
}

export interface PrefectureBuildCatalogRecipe {
  readonly level: BuildingLevel;
  readonly resources: Resources;
  readonly maintenance: number;
  readonly workforce: number;
  readonly requirements: InfrastructureRequirements;
}
